osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -265,6 +265,38 @@ private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeS
 
             foreach (int i in changes.InsertedIndices)
                 UpdateBeatmapSet(sender[i].Detach());
+
+            if (changes.DeletedIndices.Length > 0)
+            {
+                // To handle the beatmap update flow, attempt to track selection changes across delete-insert transactions.
+                // When an update occurs, the previous beatmap set is either soft or hard deleted.
+                // Check if the current selection was potentially deleted by re-querying its validity.
+                bool selectedSetMarkedDeleted = realm.Run(r => r.Find<BeatmapSetInfo>(SelectedBeatmapSet.ID))?.DeletePending != false;
+
+                if (selectedSetMarkedDeleted)
+                {
+                    // If it is no longer valid, make the bold assumption that an updated version will be available in the modified indices.
+                    // This relies on the full update operation being in a single transaction, so please don't change that.
+                    foreach (int i in changes.NewModifiedIndices)
+                    {
+                        var beatmapSetInfo = sender[i];
+
+                        foreach (var beatmapInfo in beatmapSetInfo.Beatmaps)
+                        {
+                            // Best effort matching. We can't use ID because in the update flow a new version will get its own GUID.
+                            bool selectionMatches =
+                                ((IBeatmapMetadataInfo)beatmapInfo.Metadata).Equals(SelectedBeatmapInfo.Metadata)
+                                && beatmapInfo.DifficultyName == SelectedBeatmapInfo.DifficultyName;
```

Comment:
I meant this. Isn't this true only if the difficulty also exists in the updated beatmap.

I would think if the beatmap set's metadata matches, it's OK to select any difficulty, not necessarily the one that was previously selected.

Reply:
Right. A `BeatmapSet`'s metadata is currently just a reference to the first beatmap in the set, so it gets a bit difficult to guarantee anything there.

There are other scenarios such as if the difficulty name is changed (or the title of the beatmap is changed) which will lose selection in a similar way, so if we want to be more inclusive then a fallback selection can be added. See https://github.com/ppy/osu/pull/19409/commits/24d75612e222fe3e055d1f5fabf623727dc81727.